keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1001 @@
+- An optional per-mode version label can be used to pin the policy to the version that shipped
+  with a given Kubernetes minor version (e.g. `v1.18`)
+- Dry-run of namespace updates is supported to test enforcing policy changes against existing pods.
+- Policy exemptions can be statically configured based on (requesting) user, RuntimeClass, or
+  namespace. A request meeting exemption criteria is ignored by the admission plugin.
+
+[Pod Security Standards]: https://kubernetes.io/docs/concepts/security/pod-security-standards/
+
+## Motivation
+
+Pod Security Policy is deprecated as of Kubernetes v1.21. There were numerous problems with
+PSP that lead to the decision to deprecate it, rather than promote it to GA, including:
+
+1. Policy authorization model - Policies are bound to the requesting user OR the pod’s service
+   account
+     - Granting permission to the user is intuitive, but breaks controllers
+     - Dual model weakens security
+2. Rollout challenges - PSP fails closed in the absence of policy
+     - The feature can never be enabled by default
+     - Need 100% coverage before rolling out (and no dry-run / audit mode)
+     - Leads to insufficient test coverage
+3. Inconsistent & unbounded API - Partly an aesthetic issue, but highlights lack of flexibility
+     - API has grown organically and has many internal inconsistencies (usability challenge)
+     - Unclear how to decide what should be part of PSP (e.g. fine-grained volume restrictions)
+     - Doesn’t compose well
+     - Mutation priority can be unexpected
+
+However, we still feel that Kubernetes should include a mechanism to prevent privilege escalation
+through the create-pod permission.
+
+### Goals
+
+Replace PodSecurityPolicy without compromising the ability for Kubernetes to limit privilege
+escalation out of the box. Specifically, there should be a built-in way to limit create/update pod
+permissions so they are not equivalent to root-on-node (or cluster).
+
+#### Requirements
+
+1. Validating only (i.e. no changing pods to make them comply with policy)
+2. Safe to enable in new AND upgraded clusters
+    - Dryrun policy changes and/or Audit-only mode
+3. Built-in in-tree controller
+4. Capable of supporting Windows in the future, if not in the initial release
+    - Don’t automatically break windows pods
+5. Must be responsive to Pod API evolution across versions
+6. (fuzzy) Easy to use, don’t need to be a kubernetes/security/linux expert to meet the basic objective
+7. Extensible: should work with custom policy implementations without whole-sale replacement
+    - Enable custom policies keyed off of RuntimeClassNames
+
+Nice to have:
+
+1. Exceptions or policy bindings by requesting user
+2. (fuzzy) Windows support in the initial release
+3. Admission controller is enabled by default in beta or GA phase
+4. Provide an easy migration path from PodSecurityPolicy
+5. Enforcement on pod-controller resources (i.e. things embedding a PodTemplate)
```

Comment:
There is a problem I have encountered while using the PSP.

For some components implemented in the Sidecar Injector pattern (e.g. istio proxy requires NET_ADMIN and root privileges), PodSecurityPolicy must be disabled or loosely set.

It will need to think in various directions, but at least I think the SecurityPolicy should be applied differently for each container.

Reply:
There is no way to distinguish between istio's injected sidecar and a user putting something there. So I think poking the hole for istio weakens it too much to be safe. This is why the istio cni driver was developed. It seems to strike the right balance between configuration and security IMO and doesn't need PSP's to be weakened.

Reply:
Agree with everything @kfox1111 said. The injected sidecar use case was one that we considered early on in the design process, but dismissed for these reasons. Deciding what a safe injected sidecar looks like is to implementation-specific, so if support is really needed we would just recommend implementing a custom 3rd party controller.

Reply:
agreed. when discussed more broadly, consensus has been to avoid trying to implement subdivided security policies on a pod.